atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/AtlasLockDescriptorUtils.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.immutables.value.Value;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Bytes;
+import com.palantir.lock.LockDescriptor;
+
+import okio.ByteString;
+
+public final class AtlasLockDescriptorUtils {
+    private static final byte[] ZERO_ARRAY = new byte[] {0};
+
+    private AtlasLockDescriptorUtils() {
+        // NOPE
+    }
+
+    public static List<CellReference> candidateCells(LockDescriptor lockDescriptor) {
+        Optional<TableRefAndRemainder> tableRefAndRemainder = tryParseTableRef(lockDescriptor);
+        if (!tableRefAndRemainder.isPresent()) {
+            return ImmutableList.of();
+        }
+
+        TableReference tableRef = tableRefAndRemainder.get().tableRef();
+        ByteString remainingBytes = tableRefAndRemainder.get().remainder();
+
+        int lookupFrom = 0;
```

Comment:
it feels like you're better off just getting the zeros yourself?

```java
IntStream.range(0, remainingBytes.size())
        .filter(i -> isValidZeroIndex(remainingBytes, i)
        .mapToObj(zeroIndex -> createCellFromByteString(remainingBytes, zeroIndex)
        .map(cell -> CellReference.of(tableRef, cell))
        .collect(toList());

private static boolean isValidZeroIndex(ByteString remainingBytes, int zeroIndex) {
	remainingBytes.getByte(zeroIndex) == 0 && zeroIndex > 0 && zeroIndex < remainingBytes.size() - 2
}

private static Cell createCellFromByteString(ByteString remainingBytes, int zeroIndexDelimiter) {
    byte[] row = remainingBytes.substring(0, i).toByteArray();
    byte[] col = remainingBytes.substring(i + 1, remainingBytes.size()).toByteArray();
    return Cell.create(row, col);
}
```

only because `ByteString.indexOf` basically goes through it one by one. With `IntStream` and `IntPredicate` hopefully there is no boxing/unboxing cost to pay.

Reply:
Might want to micro-benchmark this if perf is a concern. I like the readability of @felixdesouza's suggestion.

Reply:
This is much nicer, agree